Okta/OktaAuth/Tasks/OktaAuthenticate.swift

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+class AuthenticateTask: OktaAuthTask<OIDAuthState> {
+
+    private let sessionToken: String
+    
+    init(sessionToken: String, config: OktaAuthConfig, oktaAPI: OktaHttpApiProtocol) {
+        self.sessionToken = sessionToken
+        super.init(config: config, oktaAPI: oktaAPI)
+    }
+    
+    override func run(callback: @escaping (OIDAuthState?, OktaError?) -> Void) {
+        MetadataDiscovery(config: config, oktaAPI: oktaAPI).run { oidConfig, error in
+            guard let oidConfig = oidConfig else {
+                callback(nil, error)
+                return
+            }
+            
+            let codeVerifier = OIDAuthorizationRequest.generateCodeVerifier()
+            let codeChallenge = OIDAuthorizationRequest.codeChallengeS256(forVerifier: codeVerifier)
+            let state = OIDAuthorizationRequest.generateState()
```

Comment:
Probably OK for now, but we should check if the `state` was passed as an `additionalParameter` first, then default to generating a value. It is common for devs to use this param to capture before/after auth context.

Reply:
Let's do it as a separate task. This PR becomes too large.